### ISSUE TITLE
Send dm to anyone

### ISSUE THREAD

**Issue by baruKV** [RocketChat/hubot-rocketchat#136]
Hi,

at the moment I do 

```
msg.envelope.user.name = "foo"   
msg.sendDirect "test"
```

to send a dm to someone. 
Is there any other way, is it correct, is it possible to have a function like:

`sendDirect "user" "msg"` ?

**Comment by timkinnane** [comment]
What you're suggesting is possible with the current adapter, but its conceptually a little different to how you might think. Hubot doesn't send messages to users, it sends to rooms. To send a DM, you're just sending a message to the direct message room between the user and the bot. The room ID for a DM is actually just a composite of the two IDs of the users talking to each other.

So you can find the DM id for hubot to address a user, with `DM = robot.adapter.chatdriver.getDirectMessageRoomId user.name`

Then you can send to that room with `robot.adapter.chatdriver.sendMessageByRoomId msg.text, DM.room`

Note also that these are asynchronous requests, so you can't do them as consecutive commands because the room request won't return before it sends the message, you have to chain them as promises.

For querying single user details from Hubot, you can use the methods already provided in Rocket.Chat, called by Hubot like `robot.adapter.callMethod('methodName', args)`. However, I'm working on a package that's a bit more versatile and useful for bots. [Rocket-chat-bot-helpers](https://github.com/timkinnane/rocketchat-bot-helpers/) is still in development, but I'm testing it with a Hubot script for creating multiple [DM announcements](https://github.com/timkinnane/hubot-rocketchat-announcement) to an array of users, which illustrates the functionality you're looking for.

**Comment by cbrummitt** [comment]
I'm confused how to do this. 

My bot is responding to a message that contains a mention of another user. Let's call the user being mentioned the "recipient". How do I make the bot send a direct message to that recipient?

@baruKV  I'm not sure what `msg` is here. Is this the Response object? (By that I mean: is the bot running your code within a `respond` script that begins something like `robot.respond /command/i, (msg) ->`?)

@timkinnane Is `user` the User object of the person who sent the command? Is `user.name` the user name of `user` or the ID of `user`? What is your `msg` object in your example? I get an error when I try running this code:
```
robot.adapter.chatdriver.sendMessageByRoomId string_of_message_contents, robot.adapter.chatdriver.getDirectMessageRoomId(recipient_name).room
```
where `recipient_name` is the user name of the person to whom I'm trying to send a DM. In fact, 
```
robot.adapter.chatdriver.getDirectMessageRoomId(recipient)
```
produces the error 
```
TypeError: Cannot read property 'getDirectMessageRoomId' of undefined
```
both when the input `recipient` is a user name and an ID of a user.

I know that the bot can send a direct message to the person who wrote the command by simply using `sendPrivate`, thanks to commit 1203978c90fd5d5d8. But as far as I know, this function doesn't let me send a DM to a user other than the one who sent the message that triggered the `respond` command.

**Comment by timkinnane** [comment]
As to why `getDirectMessageRoomId` call isn't working the TypeError suggests that the robot.adapter.chatdriver is undefined, so it might not be available in the scope you're calling it. However, even if that weren't the case, it probably wouldn't work the way you're using it. You can't use that method synchronously as an argument to another function because it will only return immediately if it has a cached result for the room id its looking up, otherwise it has to request it from RC (calling `createDirectMessage`) and wait for a response, then it will return a promise not the room object. Have a look at my usage [here](https://github.com/timkinnane/hubot-rocketchat-announcement/blob/master/src/rocketchat-announcement.coffee#L98) - not exactly as you need it, but should provide some guide.

Regarding understanding the `msg` object, you should just do lots of console logging to examine that. Hubot can be hard to debug, I'm preparing a guide from [my own learnings](https://github.com/timkinnane/timkinnane.github.io/blob/master/articles/2016/07/Hacking%20with%20Hubot%20and%20Rocket.Chat.md#debugging-hubot), you can take a sneak ;)

**Comment by cbrummitt** [comment]
Thanks @timkinnane !

I know the user ID of the person to whom I want to send a direct message from my bot. Somewhere in the discussion of the Rocket.Chat Hubot adapter (in a thread that I can't find right now) I think I read that the room name for a direct message is simply a concatenation of the IDs of the two users (in what order I'm not sure; lexicographic?). But I can't seem to find a way to find the user ID of my bot. The bot doesn't appear in the list of users when I make the bot send `"#{Util.inspect(robot.brain.data.users)}"` to a channel.

Do you know if this method is possible (determine the bot's user ID, and then concatenate it with another user's ID)?

**Comment by timkinnane** [comment]
Yeah it is just a concatenation, that's basically what `createDirectMessage` does ([see here](https://github.com/RocketChat/Rocket.Chat/blob/00c881ec09d8a6a7df39b0b0875cfc3b71ee5e82/server/methods/createDirectMessage.coffee#L22)). Don't know if you can just give it a room id though, it might need to do some other preparation to create the DM room if that DM didn't exist already.

Don't know if the robot knows it's own user ID or if there's a RC method you'd have to call to find it. I can't stress enough you should just console.log or use node-inspector to debug the robot and msg objects to answer these, its a much faster process than taking a stab in the dark or asking here what's inside those objects.

**Comment by geekgonecrazy** [comment]
Its possible that you can get your userId from [asteroid](https://github.com/mondora/asteroid).  From a browser in the developer console you can type: `Meteor.userId()`and get it.  I'm not sure what the equivalent would be in asteroid.  A quick search through their repo yielded no results.

**Comment by cbrummitt** [comment]
Thanks @timkinnane and @geekgonecrazy. I see my user name with `Meteor.userId()`. I'm trying to figure out how to put all this together. It seems a variant of `sendDirect` (`sendDirectToUsername`?) may be needed...

**Comment by timkinnane** [comment]
Sounds good to me @cbrummitt. I'd suggest `sendDirectToUser` to accept a user object or a username to be a bit more flexible.

`SendDirectToUser` could be almost identical to `SendDirect` taking `(user, strings...)` instead of `(envelope, strings...)` because the only thing the envelope is used for is getting the room id.

**Comment by cbrummitt** [comment]
@timkinnane Do you know if you'll be able to work on this at all soon? I'm not sure how much of the technology stack we'll need to know to solve this. I don't know @[Meteor](https://www.meteor.com/) nor [asteroid](https://github.com/mondora/asteroid), and I haven't set up a way to locally develop Rocket.Chat. Ping me if you'd like to put our heads together on this.

**Comment by timkinnane** [comment]
@cbrummitt I was agreeing with your proposal that `sendDirectToUser` method would be a helpful addition to the adaptor, but I'm not intending to work on this. I've found the solutions I need and was just offering some learnings to help you with yours.
